src/coreclr/jit/lsra.cpp

```diff
@@ -11949,18 +11949,30 @@ void LinearScan::RegisterSelection::try_SPILL_COST()
                 continue;
             }
 
-            if ((recentRefPosition != nullptr) && (recentRefPosition->RegOptional() &&
-                                                   !(assignedInterval->isLocalVar && recentRefPosition->IsActualRef())))
+            RefPosition* reloadRefPosition = assignedInterval->getNextRefPosition();
+
+            if (reloadRefPosition != nullptr)
             {
-                // We do not "spillAfter" if previous (recent) refPosition was regOptional or if it
-                // is not an actual ref. In those cases, we will reload in future (next) refPosition.
-                // For such cases, consider the spill cost of next refposition.
-                // See notes in "spillInterval()".
-                RefPosition* reloadRefPosition = assignedInterval->getNextRefPosition();
-                if (reloadRefPosition != nullptr)
+                if ((recentRefPosition != nullptr) &&
+                    (recentRefPosition->RegOptional() &&
+                     !(assignedInterval->isLocalVar && recentRefPosition->IsActualRef())))
                 {
+                    // We do not "spillAfter" if previous (recent) refPosition was regOptional or if it
+                    // is not an actual ref. In those cases, we will reload in future (next) refPosition.
+                    // For such cases, consider the spill cost of next refposition.
+                    // See notes in "spillInterval()".
                     currentSpillWeight = linearScan->getWeight(reloadRefPosition);
                 }
+#ifdef TARGET_ARM64
+                else if (assignedInterval->getNextRefPosition()->needsConsecutive)
+                {
+                    // if next refposition is part of consecutive registers and there is already a register
+                    // assigned to it then do not reassign for currentRefPosition, because with that, other
+                    // registers for the next consecutive register assignment would have to be copied to
+                    // different consecutive registers since this register is busy from this point onwards.
+                    continue;
```

Comment:
Would it be possible to run out of registers if there are a number of instrinsics defining consecutive registers and later using them, and in between we don't allow spilling them?

Reply:
Yes, that was exactly on my mind and hence wanted to check if we find a case like that in CI. I think I will revert this portion.

Reply:
On a second thought, I could actually have it record the spill cost as costlier, so those registers will be preferred if no other option is available. With that we do eliminate the extra copies and the code looks much better:

main vs. latest changes

<img width="1428" alt="image" src="https://github.com/dotnet/runtime/assets/12488060/589f2bc3-8bba-45a8-b414-e5f5562bd889">


skip spilling consecutive vs. latest changes

<img width="1384" alt="image" src="https://github.com/dotnet/runtime/assets/12488060/f627466f-c1a7-46a5-a570-bc0e6076eeb7">